### ISSUE TITLE
Provide a mechanism for encoded message buffer recycling

### ISSUE THREAD

**Issue by HippoBaro** [grpc/grpc-go#6619]
### Use case(s) - what problem will this feature solve?

On their way to the wire, messages are encoded through user-provided `encoding.Codec` modules. These modules implement the following interface:

```go
// Codec defines the interface gRPC uses to encode and decode messages.  Note
// that implementations of this interface must be thread safe; a Codec's
// methods can be called from concurrent goroutines.
type Codec interface {
	// Marshal returns the wire format of v.
	Marshal(v any) ([]byte, error)
        [...]
}
```

The implementation of this interface is responsible for allocating the resulting byte slice. Once gRPC is done writing the message to the underlying transport, the reference to the slice is dropped, and garbage collection has to scavenge it. This means memory used for message encoding is never directly reused but has to go through a GC cycle before it can be allocated again.

### Proposed Solution

We propose adding an additional codec interface that complements the existing one:

 ```go
 // BufferedCodec is an optional interface Codec may implement.
// It signals the ability of the codec to use pre-existing memory
// when writing the wire format of messages.
//
// # Experimental
//
// Notice: This API is EXPERIMENTAL and may be changed or removed in a
// later release.
type BufferedCodec interface {
	// MarshalWithBuffer returns the wire format of v.
	//
	// Implementation may use a buffer from the provided buffer
	// pool when marshalling. Doing so enables memory reuse.
	MarshalWithBuffer(v any, pool SharedBufferPool) ([]byte, error)
}
 ```
 
This API takes the existing type `grpc.SharedBufferPool` that provides an allocator-like API. Codecs that can marshal into pre-existing buffers, such as `protobuf`, may implement it.

This interface should be optional because codecs need to estimate an upper bound for the size of the marshaled message to allocate a sufficiently large buffer. However, this estimate need not be accurate, as Go will reallocate buffers transparently if needed, and the GC will safely collect the original slice.

This new interface allows the allocation side to reuse buffers, but gRPC needs to return them to be recycled. We propose that gRPC return these encoded message buffers to the buffer pool once messages have been fully written to the transport and are no longer read from.

This implies adding new APIs for clients and servers, such as:

```go
// ClientEncoderBufferPool is a CallOption to provide a SharedBufferPool
// used for the purpose of encoding messages. Buffers from this pool are
// used when encoding messages and returned once they have been transmitted
// over the network to be reused.
//
// Note that a compatible encoding.Codec is needed for buffer reuse. See
// encoding.BufferedCodec for additional details. If a non-compatible codec
// is used, buffer reuse will not apply.
//
// # Experimental
//
// Notice: This API is EXPERIMENTAL and may be changed or removed in a
// later release.
func ClientEncoderBufferPool(bufferPool SharedBufferPool) CallOption {
	return EncoderBufferPoolCallOption{BufferPool: bufferPool}
}
```

and

```go
// ServerEncoderBufferPool is a ServerOption to provide a SharedBufferPool
// used for the purpose of encoding messages. Buffers from this pool are
// used when encoding messages and returned once they have been transmitted
// over the network to be reused.
//
// Note that a compatible encoding.Codec is needed for buffer reuse. See
// encoding.BufferedCodec for additional details. If a non-compatible codec
// is used, buffer reuse will not apply.
//
// # Experimental
//
// Notice: This type is EXPERIMENTAL and may be changed or removed in a
// later release.
func ServerEncoderBufferPool(bufferPool SharedBufferPool) ServerOption {
	return newFuncServerOption(func(o *serverOptions) {
		o.encoderBufferPool = bufferPool
	})
}
```

Giving users the ability to provide their implementation of buffer pools is advantageous for those that deal with specific messages of known sizes, and their implementation can be optimal. It also leverages prior art (see `grpc.WithRecvBufferPool`, for example)

### Alternatives Considered

Changes to the codec could be omitted entirely: simply returning buffers to a user-provided pool and letting them coordinate with their codec to pull from it. However, this would make it difficult for the default `proto` codec to support this feature.

Alternatively, this change could be made without any external API changes by internally creating and managing the buffer pool. This would prevent third-party codecs from benefitting from the improvement, however.

### Additional Context

We are currently running a service that uses gRPC streaming to stream (potentially) large files, in chunks, back to gRPC clients over the network. We measured that the Go allocation volume per second is roughly equal to the network throughput of the host. This creates GC cycles that introduce latency spikes and prevent us from predictably saturating the network at a reasonable CPU cost. After investigation, we have isolated the source of most of these allocations to protobuf slice creation during message serialization. 

This service demonstrated the benefit of this approach with a very significant reduction in allocation volume (>90% fewer bytes allocated per second and a 7x reduction in allocations) and much lower CPU usage (>30% less). We offer an implementation of this proposal for context and as a conversation starter: https://github.com/grpc/grpc-go/pull/6613

I should note that there was prior work on this issue here https://github.com/grpc/grpc-go/issues/2817 and here https://github.com/grpc/grpc-go/issues/2816 but these issues didn't attract much discussion.

Thank you!

**Comment by easwars** [comment]
Thanks for the detailed issue. We will get back to you shortly.

**Comment by atollena** [comment]
Gentle ping on this. We discussed this issue and the associated PR with easwars, dfawley, zasweq, joybestourous and s-matyukevich during our trip to the google office.

@dfawley expressed concerns about where buffers are released:
1. Stats handlers may keep a copy (and in fact I think you know existing usages that do, such as handlers that keep the n last RPCs for debugging and make them available via a mechanism similar to channelz). This makes the PR a breaking change.
2. I think you mentioned the tap mechanism, but from what I see only the method name is exposed there in [tap.go](https://github.com/grpc/grpc-go/blob/778860e606e35d096773133e2995fe9ecc470ba0/tap/tap.go), so perhaps you were thinking of something else.

Anything you'd like to add?

**Comment by easwars** [comment]
Apologies for the delay. I've been swamped with other things and I haven't been able to get to this yet.

**Comment by easwars** [comment]
Unfortunately I could not get to this in time and I'm going to be away from work for a while. Someone else from the team will pick this up. Thanks for your patience.

**Comment by HippoBaro** [comment]
> Stats handlers may keep a copy (and in fact I think you know existing usages that do, such as handlers that keep the n last RPCs for debugging and make them available via a mechanism similar to channelz). This makes the PR a breaking change.

Do you mean, "Stats handlers may keep a _reference_"? 

This is what I was worried about (I mentioned this in #6613). My initial thought was that doing so should not be allowed (if messages need to be preserved beyond the scope of an interceptor, they should be memcpy'd). There's no way to enforce that, though, and we'll likely make the ecosystem more brittle if we change the rule there.

There's precedent for merging these kinds of PRs and automatically turning off the feature if any interceptors are added (see #5862). Could we take a similar approach here? 

https://github.com/grpc/grpc-go/blob/e14d5831b59b0f46853d076640ae85f964164b37/dialoptions.go#L705-L707

**Comment by ash2k** [comment]
Some discussion on the same issue of stat handlers is in https://github.com/grpc/grpc-go/issues/6660.

**Comment by arvindbr8** [comment]
We have a take a step back and look at the best way to implement these buffer recycling enhancements. However, our team needs more cycles to investigate and review the changes proposed. We will keep posting updates as our investigation progresses.

**Comment by HippoBaro** [comment]
@arvindbr8 Do you have any update to share? Is there a way for me to help?

**Comment by arvindbr8** [comment]
@HippoBaro -- Our team is tracking this item internally - which includes a wider analysis and review with other implementations of gRPC. 

However, this item could not be assigned this quarter with the other high priority projects in the pipeline. Not sure if there is much to do until we pick this up. We will keep you updated.

**Comment by PapaCharlie** [comment]
Hey @arvindbr8, I'm following up on this since it's the end of the quarter. What are the odds this issue will be picked up this coming quarter? We'd see some pretty significant performance improvements on our end if this became a reality. Thanks in advance!

**Comment by dfawley** [comment]
We _might_ be doing some work on codecs next quarter depending on internal priorities.  If so, the most likely change would be to use a scatter-gather API that references the transport's buffers directly and takes ownership of the buffers.  It would probably also want to have a way to indicate that the buffers are no longer needed, as we already re-use these buffers.  The two would enable "zero"-copy in gRPC (excluding the Go std HTTP/2 framer which doesn't support it) which should improve performance on data heavy workloads for codecs that can take advantage of it (proto doesn't).

**Comment by PapaCharlie** [comment]
OK makes sense. Is there a way for us to provide any help here? For our project, we'd really like to avoid running off of a fork but we're seeing memory allocation rates on the order of multiple gigabytes per second... We even had to set up some rate limits around actually generating/sending responses to prevent the process from OOMing. We'd save a pretty significant amount of hardware with this.

**Comment by dfawley** [comment]
Apologies - now that I looked more closely at the proposal itself, I noticed that you're looking at the _sending_ of data (`Marshal`), whereas my comment was mostly about the _receiving_ side (`Unmarshal`).

For the `encoding.Codec` API (both sides), generally I'd prefer to have one style of API for recycling the buffer(s).

In both directions, the buffers being passed around might want a method on them to enable the side that produced them to re-use them.  The API needs to support buffer ownership transfer, re-use, and scatter-gather, even if any given codec (including protobuf) doesn't support those features.

Maybe something like this?

```go
type Codec interface {
	Marshal(msg any) (Buffer, error)
	Unmarshal(buf Buffer, msg any) error
}

type Buffer interface {
	Data() [][]byte
	// Free is called by the consumer of the buffer when it is done using the data buffers.
	// The producer of the buffer may then reuse it.
	Free()
}
```

It would be interesting to brainstorm other ideas and see which one is the most ergonomic.  If you're able to do that and also implement the solution that we find will be best, then that would be extremely helpful.

**Comment by PapaCharlie** [comment]
Funny you should mention that! I did do that: https://github.com/grpc/grpc-go/pull/6608. It's still rough, but it basically just extends the existing `Codec` interface to include `MarshalAppend`, which invokes the corresponding method in `google.golang.org/protobuf/proto`. This way it invokes the existing methods out of the box without explicit wiring, other than specifying the buffer pool.

The author of this issue also provided his own thoughts/approach in https://github.com/grpc/grpc-go/pull/6613. I think the right answer probably lies somewhere in between.

I've spent a decent amount of time looking into this issue and how it should be solved, so I'm very happy to help brainstorm. I however haven't spent much time actually contributing to gRPC-go so I don't want to presume my approach really matches the existing style/standards in the code.

Maybe it'd be worth having an actual meeting to talk about this? I understand that your team has a lot on its plate and may not have time to address this, but this is a big enough problem for us at LinkedIn that my team would be happy to allocate engineer power to tackle this. All you need to do is point us in the right direction!

**Comment by PapaCharlie** [comment]
So, after my meeting with @dfawley, we arrived at a proposal. The `Codec` interface can be upgraded to something like the following:
```go
type Codec interface {
	Marshal(v any) (length int, seq iter.Seq2[[]byte, error])
	ReturnBuffer([]byte)

	GetBuffer(len int) []byte
	Unmarshal(v any, length int, seq iter.Seq2[[]byte, error]) error
}
```
(this could be done by having a new `google.golang.org/grpc/encoding/v2` package)

The point of this new interface is to allow zero-copy codecs to exist, and allowing those codecs to (optionally) reuse buffers. In this case, upgrading the existing gRPC codec to this interface is trivial. Both `ReturnBuffer` and `GetBuffer` could be made optional to keep the interface simple. The server implementation can iterate over the returned sequence of buffers without ever fully materializing the entire buffer.

The `iter.Seq2[[]byte, error]` type doesn't exist for older versions of go but is simply `func(yield func([]byte, error) bool)`, and can be upgraded when the oldest supported version of `grpc-go` is go1.22.

We want to open the discussion on this before jumping in head first, please let us know what you think!

**Comment by vmg** [comment]
This works for us at [Vitess](https://github.com/vitessio/vitess/). We're very interested on having a scatter-gather kind of API, particularly for serialization. Just thumbs up overall. Anything I can do to help with the implementation and/or early testing?

**Comment by PapaCharlie** [comment]
Hey @vmg good to hear! Can you give me some more insight on how you'd use this? Just trying to get a better feel for what should be prioritized first. There's some interesting implications with what I've proposed. For example, [some of the public APIs](https://github.com/grpc/grpc-go/blob/55cd7a68b3c18a0f76ea9c1be37221a5b901a798/internal/transport/transport.go#L728) may have to be changed to accept the new `iter.Seq2[[]byte, error]` type, among other things

**Comment by jzelinskie** [comment]
We're also interested in this for SpiceDB. Most of our allocations are coming from gRPC messages, so anything to reuse those helps our overall scalability and latency. I'm not sure we have strong opinions on the hook to provide the data slice, just that the slices can be reused. A simple sync.Pool implementation would be heaps (pun intended) of help.

**Comment by vmg** [comment]
@PapaCharlie: for Vitess, we happen to often work with very large packets coming from the MySQL binary protocol which we must make available through the distributed system via GRPC. Although the default ProtoBuf Go library does not support incremental/chunked serialization, I maintain an alternative implementation (https://github.com/planetscale/vtprotobuf/) that Vitess and many other OSS projects use. Being able to serialize as an `iter.Seq2[[]byte, error]` would be a game changer for these large ProtoBuf messages that contain binary blobs inside of them.

**Comment by PapaCharlie** [comment]
Hey everyone, circling back on this after actually implementing a lot of this. Here's the interface that we landed on:
```go
package encoding

type CodecV2 interface {
	Marshal(v any) (out mem.BufferSlice, err error)
	Unmarshal(data mem.BufferSlice, v any) error
	Name() string
}

...

package mem

type BufferSlice []*Buffer

type Buffer struct {
	data []byte
	refs *atomic.Int32
	free func([]byte)
}
```

These `Buffer`s will be used to read off of the wire, and come from a shared buffer pool. Once the `CodecV2` is done with the buffers given to it by `Unmarshal`, it can call `Free` to return it back to the server's pool. It can hold on onto those buffers as long as it wants (e.g. to reference them directly instead of making a direct copy). Conversely, the `Buffer`s coming out of `Marshal` can be sampled from the same backing pool if needed (there are some improved implementations of buffer pools in the `mem` package). Let us know what you think and if this works for you!

**Comment by PapaCharlie** [comment]
Hey @vmg, what are your thoughts on this? Got some thumbs up from the other folks on this issue but since you maintain vtprotobuf, your generated code will likely be able to make the best use of this

**Comment by vmg** [comment]
This seems sensible to me! To be fair, there may be a sharp corner hidden here which is not obvious by looking at the API. My homeboy @frouioui is going to try your CodecV2 branch next week to attempt a Vitess optimization which we've been planning for a while. I think he'll have more meaningful feedback very soon.

**Comment by GiedriusS** [comment]
@vmg maybe there are some updates?

**Comment by jzelinskie** [comment]
I took a stab at implementing a vtprotobuf Codec V2 now that #7356 was merged.

Does this look right?

**Comment by vmg** [comment]
@jzelinskie: I'm not quite sure! Could you please open a PR on the upstream repository and we can review and discuss it?

**Comment by coxley** [comment]
Are there benchmarks for the buffer recycling available? 

I'm consistently getting better results with the original codec. I can't tell if my tests are subtly wrong, or if using `go test -bench` is hard to accurately assess pooling.

Full code in gist: https://gist.github.com/coxley/3c3eab255f5a34cf43c33491f3c21703

**With CodecV2**:
```go
➜ go test -bench=. -benchmem -run=Benchmark
goos: darwin
goarch: arm64
pkg: github.com/TriggerMail/go-common/grpc/codec
cpu: Apple M1 Pro
BenchmarkCodec/vtproto/sm-marshal-10            14379588                79.34 ns/op           56 B/op          3 allocs/op
BenchmarkCodec/vtproto/sm-unmarshal-10          25870966                46.23 ns/op           88 B/op          2 allocs/op
BenchmarkCodec/vtproto/md-marshal-10             1742565               668.9 ns/op          4210 B/op          5 allocs/op
BenchmarkCodec/vtproto/md-unmarshal-10           4537872               267.6 ns/op          2128 B/op          2 allocs/op
BenchmarkCodec/vtproto/lg-marshal-10              364785              3224 ns/op           32922 B/op          5 allocs/op
BenchmarkCodec/vtproto/lg-unmarshal-10            450477              2689 ns/op           32848 B/op          2 allocs/op
BenchmarkCodec/vanilla/sm-marshal-10            10000597               124.9 ns/op            56 B/op          3 allocs/op
BenchmarkCodec/vanilla/sm-unmarshal-10          11265732               106.2 ns/op            72 B/op          2 allocs/op
BenchmarkCodec/vanilla/md-marshal-10             1323927               887.1 ns/op          4210 B/op          5 allocs/op
BenchmarkCodec/vanilla/md-unmarshal-10           2500578               477.9 ns/op          2112 B/op          2 allocs/op
BenchmarkCodec/vanilla/lg-marshal-10              214489              5633 ns/op           32925 B/op          5 allocs/op
BenchmarkCodec/vanilla/lg-unmarshal-10            240549              4930 ns/op           32832 B/op          2 allocs/op
PASS
ok      github.com/TriggerMail/go-common/grpc/codec     17.842s
```

**With Old**:

```
➜ go test -bench=. -benchmem -run=Benchmark
goos: darwin
goarch: arm64
pkg: github.com/TriggerMail/go-common/grpc/codec
cpu: Apple M1 Pro
BenchmarkCodec/vtproto/sm-marshal-10            60273174                20.61 ns/op           16 B/op          1 allocs/op
BenchmarkCodec/vtproto/sm-unmarshal-10          28437141                42.37 ns/op           88 B/op          2 allocs/op
BenchmarkCodec/vtproto/md-marshal-10             4884146               280.5 ns/op          2304 B/op          1 allocs/op
BenchmarkCodec/vtproto/md-unmarshal-10           4638386               265.6 ns/op          2128 B/op          2 allocs/op
BenchmarkCodec/vtproto/lg-marshal-10              447200              2896 ns/op           32768 B/op          1 allocs/op
BenchmarkCodec/vtproto/lg-unmarshal-10            416497              2615 ns/op           32848 B/op          2 allocs/op
BenchmarkCodec/vanilla/sm-marshal-10            24079723                49.58 ns/op           16 B/op          1 allocs/op
BenchmarkCodec/vanilla/sm-unmarshal-10          13298110                91.09 ns/op           72 B/op          2 allocs/op
BenchmarkCodec/vanilla/md-marshal-10             2805237               428.7 ns/op          2304 B/op          1 allocs/op
BenchmarkCodec/vanilla/md-unmarshal-10           2647936               458.7 ns/op          2112 B/op          2 allocs/op
BenchmarkCodec/vanilla/lg-marshal-10              234085              5282 ns/op           32768 B/op          1 allocs/op
BenchmarkCodec/vanilla/lg-unmarshal-10            231348              5271 ns/op           32832 B/op          2 allocs/op
PASS
ok      github.com/TriggerMail/go-common/grpc/codec     17.592s
```

**Comment by coxley** [comment]
Aha, found it. There's an edge case when `len(mem.BufferSlice) == 1` that causes values to never get sent back to the pool. Will submit a PR.

**Comment by coxley** [comment]
Addressed in #7571

**Comment by arjan-bal** [comment]
Closing since the changes were released in v1.66.0.
